code/iaas/logic/src/main/java/io/cattle/platform/process/host/HostRemove.java

```diff
@@ -0,0 +1,116 @@
+package io.cattle.platform.process.host;
+
+import io.cattle.iaas.cluster.process.lock.ClusterLock;
+import io.cattle.iaas.cluster.service.ClusterManager;
+import io.cattle.platform.core.constants.AgentConstants;
+import io.cattle.platform.core.constants.ClusterConstants;
+import io.cattle.platform.core.constants.CommonStatesConstants;
+import io.cattle.platform.core.dao.ClusterHostMapDao;
+import io.cattle.platform.core.model.Agent;
+import io.cattle.platform.core.model.Host;
+import io.cattle.platform.core.model.tables.records.ClusterHostMapRecord;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.engine.process.impl.ProcessCancelException;
+import io.cattle.platform.lock.LockCallbackNoReturn;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.process.StandardProcess;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.process.base.AbstractDefaultProcessHandler;
+import io.cattle.platform.process.common.util.ProcessUtils;
+import io.cattle.platform.process.lock.AgentCleanupLock;
+import io.cattle.platform.process.util.ProcessHelpers;
+
+import java.util.HashMap;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+@Named
+public class HostRemove  extends AbstractDefaultProcessHandler{
```

Comment:
Do you think it would make sense to make it a pre-listener to the host.remove? So it would be easier to separate cluster related logic if we ever decide to move swarm to its separate project.